File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {

        let map= new Map();

        let current= head;

        while(current){
            if(map.has(current))
            {
                return true;
            }
            else{
                map.set(current,true)
            }
            current= current.next;
        }

        return false;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
         Map<Node, Node> map = new HashMap<>();
         Node current = head;
         Node dummy= new Node(0);
         Node tail=dummy;
         while(current != null)
         {
            Node copynode= new Node(current.val);
            map.put(current,copynode);
            tail.next=copynode;
            tail=copynode;
            current= current.next;
         }
         current= head;
         Node copyCurrent = dummy.next;  
         while(current != null)
         {
            
            copyCurrent.random= map.get(current.random);
            
            current= current.next;
            copyCurrent= copyCurrent.next;
         }

         return dummy.next;

    }
}
